extensions/amp-access-scroll/0.1/scroll-impl.js

```diff
@@ -364,3 +372,72 @@ function addAnalytics(ampdoc, vendorConfig) {
   // Append
   ampdoc.getBody().appendChild(analyticsElem);
 }
+
+class PositionTracker {
+  /**
+   * Sets up tracking of paragraphs in document.
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!../../amp-access/0.1/amp-access-source.AccessSource} accessSource
+   */
+  constructor(ampdoc, accessSource) {
+    this.ampdoc_ = ampdoc;
+    this.lastTopSnippet_ = null;
+    this.accessSource_ = accessSource;
+    this.debouncedUpdateLastRead_ = debounce(
+        ampdoc.win,
+        this.updateLastRead.bind(this),
+        1000
+    );
+
+    installPositionObserverServiceForDoc(ampdoc);
+    const positionObserver = getServiceForDoc(ampdoc, 'position-observer');
+
+    const paragraphs = ampdoc.getBody().getElementsByTagName('p');
+    for (let i = 0; i < paragraphs.length; i++) {
+      const p = paragraphs[i];
+      positionObserver.observe(p, PositionObserverFidelity.HIGH, position =>
+        this.checkPosition(p, position)
+      );
+    }
+  }
+
+  /**
+   * Checks position of a paragraph and decides whether to update server.
+   * @param {Element} paragraph
+   * @param {!../../../src/service/position-observer/position-observer-worker.PositionInViewportEntryDef} position PositionObserver entry
+   */
+  checkPosition(paragraph, position) {
+    // Ignore if position is not above viewport
+    if (position.relativePos !== RelativePositions.TOP) {
+      return;
+    }
+
+    // Update the snippet representing latest paragraph above viewport
+    const prevSnippet = this.lastTopSnippet_;
+    this.lastTopSnippet_ = paragraph./*OK*/innerText.substring(0, 50);
+    if (prevSnippet && prevSnippet !== this.lastTopSnippet_) {
+      // Send update to server if the latest snippet has changed
+      this.debouncedUpdateLastRead_(prevSnippet);
```

Comment:
Hmm this isn't very good naming. I renamed to `lastIntersectingSnippet` since the paragraph that's recognized as `RelativePositions.TOP` is the one that intersects with the top of the viewport. So then the previous snippet that's sent to the server is the latest paragraph that's now no longer in the viewport at all.